spec/integration/rack_app/middleware_spec.rb

```diff
@@ -597,15 +599,15 @@ def handle(req, res)
       expect(last_response.headers).to_not have_key("X-API-Version")
     end
 
-    it "uses Rack middleware for other paths" do
-      get "/foo"
+    it "does not use Rack middleware for other paths" do
+      get "/__not_found__"
 
       expect(last_response.status).to be(404)
-      expect(last_response.headers["X-Identifier-Root"]).to eq("true")
-      expect(last_response.headers).to have_key("X-Elapsed")
-      expect(last_response.headers).to_not have_key("X-Auth-User-ID")
-      expect(last_response.headers).to_not have_key("X-API-Rate-Limit-Quota")
-      expect(last_response.headers).to_not have_key("X-API-Version")
+      expect(last_response.headers).not_to have_key("X-Identifier-Root")
+      expect(last_response.headers).not_to have_key("X-Elapsed")
+      expect(last_response.headers).not_to have_key("X-Auth-User-ID")
+      expect(last_response.headers).not_to have_key("X-API-Rate-Limit-Quota")
+      expect(last_response.headers).not_to have_key("X-API-Version")
```

Comment:
This is one behavioural change worth noting — a request for a not found URL is no longer going to have all the headers from other middleware preserved in the final response.

I think this is actually a more reasonable position — errors responses like this should remain as simple and as independent as possible.